Add resources azurerm_virtual_network_gateway and azurerm_virtual_network_gateway_connection

dominik-lekse opened the pull request:
This pull request adds the resources `azurerm_virtual_network_gateway` and `azurerm_virtual_network_gateway_connection` which can be used to manage [Azure VPN Gateways and Connections](https://docs.microsoft.com/en-us/azure/vpn-gateway/vpn-gateway-about-vpngateways).

It has been migrated from the pull request https://github.com/hashicorp/terraform/pull/13886 in the original terraform repository after provider separation.

dominik-lekse commented:
@tombuildsstuff Please consider that I commented your review comments in the original pull request #13886 and not in this one. Most comments have been addressed, however the question how to properly handle the subnet deletion issue remains open.

veeresh1982 commented:
@dominik-lekse : Just wanted to check any ETA when will this be pushed to the master branch. We have a requirement and want to use this new resource feature. Is it OK now to push the current changes to the master and improve upon the subnet deletion issue later as a new PR ? 
Let us know your thoughts!

tombuildsstuff commented:
Hey @veeresh1982 

> Just wanted to check any ETA when will this be pushed to the master branch. We have a requirement and want to use this new resource feature. Is it OK now to push the current changes to the master and improve upon the subnet deletion issue later as a new PR ?

Unfortunately we're in a holding pattern with this PR until the underlying Azure API is fixed - as the API's returning that the Virtual Network Gateway's been deleted when it's not. There's an [Github issue tracking the bug here](https://github.com/Azure/azure-rest-api-specs/issues/1233) - and I've [commented asking for an update](https://github.com/Azure/azure-rest-api-specs/issues/1233#issuecomment-318011822) / rough timeline for this being fixed.

Whilst we could potentially merge this in now, it'd give end users random failures when applying their Terraform configs - which isn't an optimal user experience. In the interim - you should be able to provision a Virtual Network Gateway using [the `azurerm_template_deployment` resource](https://www.terraform.io/docs/providers/azurerm/r/template_deployment.html) - which (whilst isn't ideal) does allow for this resource to be provisioned via Terraform until this is natively supported.

Thanks!

dominik-lekse commented:
Hi @tombuildsstuff

Thanks for the status update on this. Besides the subnet deletion issue, did you have a chance to take a look at the changes which have been applied in accordance to your last review?

KandrewC commented:
I'm taking a look at this PR right now and trying to build the provider, how would I pull it down and run it locally?

tombuildsstuff commented:
Hey @KandrewC

> I'm taking a look at this PR right now and trying to build the provider, how would I pull it down and run it locally?

There's [instructions on the README](https://github.com/terraform-providers/terraform-provider-azurerm#building-the-provider) for how to build and run this - however bear in mind PR's aren't fully tested and may not necessarily be 100% bug free. As such in the interim we'd recommend using the `azurerm_template_deployment` resource to provision Virtual Network Gateway's at this point in time for Production usage.

Thanks!

dominik-lekse commented on azurerm/resource_arm_virtual_network_gateway.go:
```diff
@@ -0,0 +1,686 @@
+		if !ok {
+			return nil, err
+		}
+	}
```
I thought of this as well and think that your proposal is the better solution. As far as I see, Terraform currently supports validation for individual fields as part of the schema ("SchemaValidateFunc is a function used to validate a single field in the schema") but not validation for fields which in turn depends on the value of other fields of the same resource instance. In case this is possible, it would be great if you could point to an example. If not this could be an improvement to terraform core, i.e. to extend SchemaValidateFunc by providing an instance of the resource to the validator function.

tombuildsstuff replied:
Got you - so unfortunately that's not possible right now, however there's an issue open on the main repository to be able to access other properties at validation time.

dominik-lekse replied:
I found https://github.com/hashicorp/terraform/issues/12021 and an older pull request which has been closed https://github.com/hashicorp/terraform/pull/4348. Reads like there is something blocking to implement resource level validation.

dominik-lekse commented on azurerm/resource_arm_virtual_network_gateway.go:
```diff
@@ -0,0 +1,686 @@
+		if !ok {
+			return nil, err
+		}
+	}
```
See above.

dominik-lekse commented on azurerm/resource_arm_virtual_network_gateway_connection.go:
```diff
@@ -0,0 +1,350 @@
+			"shared_key": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Sensitive: true,
```
As the shared key is mandatory and not computed by the Azure API, I suggest to keep this a non-computed field.

tombuildsstuff replied:
Sounds sensible, however given it's mandatory - can we make this Required instead of Optional?

dominik-lekse replied:
I made it optional, since Site-to-Site and VNet-to-VNet connections require the key but not ExpressRoute connections. To clarify, I added to the field in the documentation "A key must be provided if a Site-to-Site or VNet-to-VNet connection is created whereas ExpressRoute connections do not need a shared key."

A more elegant way for validation would be if individual schema validator functions would have access to the entire resource to achieve "conditional mandatory" fields (see comments above).

tombuildsstuff replied:
> I made it optional, since Site-to-Site and VNet-to-VNet connections require the key but not ExpressRoute connections. To clarify, I added to the field in the documentation "A key must be provided if a Site-to-Site or VNet-to-VNet connection is created whereas ExpressRoute connections do not need a shared key."

Awesome - thanks :)

> A more elegant way for validation would be if individual schema validator functions would have access to the entire resource to achieve "conditional mandatory" fields (see comments above).

Agreed - I believe there's an issue open tracking this request on the main repository

dominik-lekse commented:
@tombuildsstuff Thanks for the review, in the the latest changes I have addressed most of your comments.

Regarding the subnet deletion issue, I have another idea for a workaround which I wanted to discuss before implementing: In `resourceArmVirtualNetworkGatewayDelete` after deleting the gateway resource we could also delete the associated gateway subnet and recreate it immediately afterwards with the same parameters. This should not interfer with other resources since there is a one-to-one-relationship between the gateway subnet and the gateway. Further, we could take out the special gateway subnet retry deletion handling from `resourceArmSubnetDelete`. Do you think this could be a workaround to allow this pull request to be merged before Microsoft finds and fixes the problem?

StephenThomson commented:
Looks great. I'm looking forward to it being merged :+1:

dominik-lekse commented:
@tombuildsstuff As @MohitGargVpn [announced](https://github.com/Azure/azure-rest-api-specs/issues/1233#issuecomment-322889187) that the fix for the subnet deletion issue will be released in the Azure Api soon, I will remove the workaround from the subnet resource in this pull request and we wait for the acceptance tests to turn green.

mynkow commented:
Is there a beta where I can test this?

nbering commented:
@mynkow You could checkout the pull request and build it yourself. Not trivial by any means, but not terribly difficult, either.

There are build instructions in the project readme, and and little bit about pulling in custom your own plugin binaries in the Terraform docs under [Running Terraform in Automation](https://www.terraform.io/guides/running-terraform-in-automation.html#pre-installed-plugins). Should really be it's own section and cross-linked from plugin development, but if I have time I'll raise an issue about that.

fstuck37 commented:
Hi All,
Any update on when this will be available?
Working with the null_resource and the az CLI as of now but would much rather have both virtual_network_gateway and azurerm_virtual_network_gateway_connection available.

Thanks and your work on this is very much appreciated,
Fred

tombuildsstuff commented:
@fstuck37 we're still waiting on Microsoft to fix the API [which is scheduled for the end of next week](https://github.com/Azure/azure-rest-api-specs/issues/1233#issuecomment-322889187), we'll take another look after that :)

dominik-lekse commented on azurerm/resource_arm_subnet.go:
```diff
@@ -236,5 +238,48 @@ func resourceArmSubnetDelete(d *schema.ResourceData, meta interface{}) error {
+			// Retry deletion of gateway subnet if provisioning state is failed
+			if *resp.SubnetPropertiesFormat.ProvisioningState == "Failed" {
+				log.Printf("[DEBUG] Retry deleting Gateway Subnet %s/%s after failed provisioning state.", vnetName, name)
+				subnetClient.Delete(resGroup, vnetName, name, make(chan struct{}))
```
This workaround should not be relevant anymore.

dominik-lekse commented:
@tombuildsstuff I have removed the workaround for the subnet deletion issue from this pull request. Also, I updated the acceptance tests according the latest test convention, i.e. parameterized locations.

The bad news is, that the subnet deletion issue still exists.

```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./azurerm -v -run=TestAccAzureRMVirtualNetworkGateway_basic -timeout 120m
=== RUN   TestAccAzureRMVirtualNetworkGateway_basic
--- FAIL: TestAccAzureRMVirtualNetworkGateway_basic (1804.44s)
	testing.go:494: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.
		
		Error: Error applying: 1 error(s) occurred:
		
		* azurerm_subnet.test (destroy): 1 error(s) occurred:
		
		* azurerm_subnet.test: network.SubnetsClient#Delete: Failure sending request: StatusCode=200 -- Original Error: Long running operation terminated with status 'Failed': Code="InternalServerError" Message="An error occurred."
		
		State: azurerm_resource_group.test:
		  ID = /subscriptions/81046569-f43e-4139-a2b4-0880e5fdd6cd/resourceGroups/acctestRG-9044211118209098620
		  location = westeurope
		  name = acctestRG-9044211118209098620
		  tags.% = 0
		azurerm_subnet.test:
		  ID = /subscriptions/81046569-f43e-4139-a2b4-0880e5fdd6cd/resourceGroups/acctestRG-9044211118209098620/providers/Microsoft.Network/virtualNetworks/test-9044211118209098620/subnets/GatewaySubnet
		  address_prefix = 10.0.1.0/24
		  ip_configurations.# = 1
		  ip_configurations.575838843 = /subscriptions/81046569-f43e-4139-a2b4-0880e5fdd6cd/resourceGroups/acctestRG-9044211118209098620/providers/Microsoft.Network/virtualNetworkGateways/test-9044211118209098620/ipConfigurations/vnetGatewayConfig
		  name = GatewaySubnet
		  network_security_group_id = 
		  resource_group_name = acctestRG-9044211118209098620
		  route_table_id = 
		  virtual_network_name = test-9044211118209098620
		
		  Dependencies:
		    azurerm_resource_group.test
		    azurerm_virtual_network.test
		azurerm_virtual_network.test:
		  ID = /subscriptions/81046569-f43e-4139-a2b4-0880e5fdd6cd/resourceGroups/acctestRG-9044211118209098620/providers/Microsoft.Network/virtualNetworks/test-9044211118209098620
		  address_space.# = 1
		  address_space.0 = 10.0.0.0/16
		  dns_servers.# = 0
		  location = westeurope
		  name = test-9044211118209098620
		  resource_group_name = acctestRG-9044211118209098620
		  subnet.# = 1
		  subnet.2563334686.address_prefix = 10.0.1.0/24
		  subnet.2563334686.name = GatewaySubnet
		  subnet.2563334686.security_group = 
		  tags.% = 0
		
		  Dependencies:
		    azurerm_resource_group.test
FAIL
exit status 1
FAIL	github.com/terraform-providers/terraform-provider-azurerm/azurerm	1804.472s
make: *** [testacc] Error 1
```

tombuildsstuff commented:
👋 @dominik-lekse 

> I have removed the workaround for the subnet deletion issue from this pull request. Also, I updated the acceptance tests according the latest test convention, i.e. parameterized locations.

Awesome - thanks for doing that :)

> The bad news is, that the subnet deletion issue still exists.

😞 that's frustrating - I've posted a comment on the other issue to find out if the fix has been released or not (I'm hoping it's just not released yet). Once we've determined that - we can determine how to proceed with this.

Thanks!

mikedbrook commented:
I could really use this function right now as my organisation wants to provision all the vnets, peering and gateways using terraform

tombuildsstuff commented:
I've just finished re-running the tests - but this API still appears to be broken:

```
$ acctests azurerm TestAccAzureRMVirtualNetworkGatewayConnection_
=== RUN   TestAccAzureRMVirtualNetworkGatewayConnection_importSiteToSite
--- FAIL: TestAccAzureRMVirtualNetworkGatewayConnection_importSiteToSite (1649.00s)
	testing.go:494: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Error applying: 1 error(s) occurred:

		* azurerm_subnet.test (destroy): 1 error(s) occurred:

		* azurerm_subnet.test: network.SubnetsClient#Delete: Failure sending request: StatusCode=200 -- Original Error: Long running operation terminated with status 'Failed': Code="InternalServerError" Message="An error occurred."
```

I've enquired as to the fix's roll-out here: https://github.com/Azure/azure-rest-api-specs/issues/1233#issuecomment-332470830

YuriyKischenko commented:
Hi all,

I was using this pull request some time ago to create virtual network gateway with subnet name **gatewaySubnet**, but the latest version is requiring me to rename it to **GatewaySubnet**, which does not makes sense as it's perfectly working with the old name. The problem is that it's not possible to rename subnet without recreating virtual network gateway which would take a lot of time.

Would it be possible to have case insensitive name comparison for subnet name?

dominik-lekse commented:
@YuriyKischenko Thanks for the feedback. I have added this in the recent commit. Could you test it for your case?

dominik-lekse commented:
@tombuildsstuff Do you think we should consider pulling the workarounds for the subnet deletion issue out of the drawer? The most simple one is just waiting a fixed amount of time after a virtual network has been deleted and not releasing a lock until the timer is up.

Once the Azure team has managed to fix the bug in the API, we just remove it from the code in a separate pull request. We should consider that the fix is blocker for this feature since around 6 months.

tombuildsstuff commented:
@dominik-lekse possibly - I was thinking the same thing but I'm concerned this could become whack-a-mole for issues in other resources. Given we're a couple of weeks away from the fix being deployed (2017-11-17) - what do you think about waiting for that and re-testing and then if it's still an issue then bring back the workarounds?

tombuildsstuff commented:
@dominik-lekse looking into this a little more, it's actually been over a year since the original PR was opened (October 2016) - I've just reached out to MS to see if they can confirm this is fixed in a staging environment before it's rolled out fully

kmcquade commented:
@tombuildsstuff just to double check - you said this fix is being deployed on Nov 17 (this upcoming friday). I'm assuming you mean that the fix will be for terraform-provider-azurerm, not just the Microsoft API itself. If you do mean the terraform-provider-azurerm will be fixed, are you still on track for completing that this Friday?

tombuildsstuff commented:
> @tombuildsstuff just to double check - you said this fix is being deployed on Nov 17 (this upcoming friday). I'm assuming you mean that the fix will be for terraform-provider-azurerm, not just the Microsoft API itself. If you do mean the terraform-provider-azurerm will be fixed, are you still on track for completing that this Friday?

@kmcquade this is a fix to the API within Azure - and not within the AzureRM Terraform Provider. If it's still broken after this we'll look into workarounds; @dominik-lekse has previously implemented a fix for this (but we've been trying to get the API fixed, rather than working around it).

tombuildsstuff commented:
@dominik-lekse 👋 I've just re-run the tests and it looks like the API's still broken :(

Would you be able to add back in the Subnet deletion workaround as you previously suggested? I believe that should allow the tests to pass which would allow us to get this merged in - sorry to mess you around here!

Thanks!

retheshnair commented:
@dominik-lekse  Can you please get a work around for this issue . We are looking forward  for this feature .

dominik-lekse commented:
Hi @tombuildsstuff, the Azure API internals remain a mystery.

Unfortunately, I cannot attribute time this week to this project, testing is a real time killer on this pull request. If you or anybody else is willing to finish, I am happy to grant write permissions on the branch of this pull request. As an alternative, we could also move this branch into the terraform-provider-azurerm repo for collaborating.

alanvanwyk commented:
Any news guys?

tombuildsstuff commented:
@dominik-lekse no worries - I'll see if I can progress with this next week

kmcquade commented:
@tombuildsstuff Any progress on Virtual Network Gateways support? If I'm not mistaken, the subnet thing shouldn't be an issue anymore. 

App Gateways and Local Gateways are covered (was happy to see BGP support on the recent release) but I was really hoping to see VNet Gateways supported soon so we can get VPN set up between AWS and Azure. I'm sure I'm not the only one who wants to be able to do this, as I've seen several repos demonstrating AWS<=>GCP VPNs. lmk

Thanks for all the great work. Respect

kmcquade commented:
@tombuildsstuff - estimated time of completion in Jan possibly?

kmcquade commented:
Created a github repo to provide an example of using azurerm_template_deployment to deploy a P2S VPN, since we are still waiting on this pull request. Comments welcome. Hope this helps someone else who is looking for a solution while we wait for the feature to be integrated with the azurerm provider plugin :)

Link to repo: https://github.com/kmcquade/terraform-azure-p2s-vpn

alanvanwyk commented:
Thanks - will look and feedback.
I’ve been running a powershell script post build - would prefer using tf natively as far as possible

On 30 Dec 2017, 02:27 +0000, Kinnaird McQuade <notifications@github.com>, wrote:
> Created a github repo to provide an example of using azurerm_template_deployment to deploy a P2S VPN, since we are still waiting on this pull request. Comments welcome. Hope this helps someone else who is looking for a solution while we wait for the feature to be integrated with the azurerm provider plugin :)
> Link to repo: https://github.com/kmcquade/terraform-azure-p2s-vpn
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

russellds commented:
I think it is time to merge this PR. It is now working correctly in US East. I haven't had a chance to test in other regions however if needed we could put a warning about it not working in all regions. I would be better to have this resource with issues, than not having the resource at all.

Additionally, having the VNet Gateway connection resource would be valuable as well.

tombuildsstuff commented:
@kmcquade apologies for the delayed response here

I'd been hoping to take a look into this in December but unfortunately the updates to Application Gateways have taken longer than planned. In the interim Microsoft have commented that this _should_ be fixed now - and I'll be picking this up once the work to Application Gateways have been completed.

Thanks!

tombuildsstuff commented:
👋 hey folks!

So I've got some good news - I've just run the acceptance tests and most of them are now passing; so it appears the API fixes have rolled out.

<img width="750" alt="screen shot 2018-01-05 at 11 50 52" src="https://user-images.githubusercontent.com/666005/34608177-f7ad0d2c-f20e-11e7-82f1-7fc9beea3a59.png">

The test failure in the above image is due to an HTTP 429 which isn't handled in v11 of the Azure SDK - however I believe should be in v12.X:

```
* azurerm_virtual_network_gateway.test: network.VirtualNetworkGatewaysClient#Delete: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="RetryableError" Message="A retryable error occurred." Details=[{"code":"RetryableErrorDueToAnotherOperation","message":"Operation DeleteVirtualNetworkGatewayOperation (2079ad8c-da79-4e26-8955-f678b25314e7) is updating resource *******. The call can be retried in 11 seconds."}]
```

As such, I believe this should be fixed in the next version of the Azure SDK for Go (v12.X) - which we're planning on upgrading too later this month. Once that's merged in we should be able to rebase this + proceed with this PR 👍 

Thanks!

dominik-lekse commented:
Great, thanks for following up on this. Sounds like the story will find its end.

I have some changes pending on this pull request which are mainly additional error/nil checks and will find time next week to commit them.

dominik-lekse commented:
@tombuildsstuff Did you check already that the HTTP 429 is handled correctly by the v12 SDK? If you perform the upgrade to v12 for the networking resources, I can work on migrating the vnet gateway resources to the new SDK.

dominik-lekse commented:
@tombuildsstuff With the recent migration to the v12 networking sdk in #738, I also migrated the resources of this pull request.

Finally the good news is, that the eventual consistency bug in the Azure API is fixed and that the v12 SDK handles the return code in the correct way. Up to now, I only run the test `TestAccAzureRMVirtualNetworkGateway_basic` targeting the West Europe region which finished without the errors we have seen in the past.

```
=== RUN   TestAccAzureRMVirtualNetworkGateway_basic
--- PASS: TestAccAzureRMVirtualNetworkGateway_basic (2107.51s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	2107.532s
```

Running the other tests is pending since each test takes ~45 minutes for the virtual network gateway to create. Could you take over and run these tests on the HashiCorp CI?

tombuildsstuff commented:
@dominik-lekse awesome - that's great news! Sure - I'll kick off the test suite now :)

Thanks!

tombuildsstuff commented:
@dominik-lekse all the tests are passing 🎉 I'll take another look through this shortly, but that's really exciting :)

<img width="501" alt="screen shot 2018-01-24 at 15 44 04" src="https://user-images.githubusercontent.com/666005/35338460-4a762a70-011e-11e8-89cd-09cde985621e.png">

dominik-lekse commented:
Sounds great, of course I expected this ;)

Since the original development of these provider resources, Azure has added support for custom IPsec/IKE policies (see https://docs.microsoft.com/en-us/azure/vpn-gateway/vpn-gateway-ipsecikepolicy-rm-powershell). I suggest to add them in a separate pull request for the sake of closing this one.

Adding the custom policies to the provider would be great since this is up to now neither supported by the Portal nor by the official Azure CLI.

dominik-lekse commented:
Resolved the conflicts with the previous commit. Ready to merge in my view.

tombuildsstuff commented:
Virtual Network Gateway & Virtual Network Gateway Connection tests pass:

<img width="486" alt="screen shot 2018-01-26 at 12 22 33" src="https://user-images.githubusercontent.com/666005/35439673-a429ff94-0293-11e8-87c5-197459146464.png">

dominik-lekse commented:
@tombuildsstuff :clap:

Thanks for the improvements and merging the resources.
